# PersonalUtilities

This is a repository for several scripts used in a few studies.

* **diversification_rate_sliding_window.middle.py** was customized for [Zhang SD, Jin JJ, et al. 2017. Diversification of Rosaceae since the Late Cretaceous based on plastid phylogenomics. New Phytologist.](https://nph.onlinelibrary.wiley.com/doi/full/10.1111/nph.14461)

* **TreePar.plot.R** was customized for [Zhang SD, Jin JJ, et al. 2017. Diversification of Rosaceae since the Late Cretaceous based on plastid phylogenomics. New Phytologist.](https://nph.onlinelibrary.wiley.com/doi/full/10.1111/nph.14461)

* **concatenate_fasta.py** was first used in [Zhang R, Wang YH, Jin JJ, et al. 2020. Exploration of plastid phylogenomic conflict yields new insights into the deep relationships of Leguminosae. Systematic Biology](https://academic.oup.com/sysbio/article/69/4/613/5739458)

* **get_annotated_regions_from_gb.py** was first used in [Zhang R, Wang YH, Jin JJ, et al. 2020. Exploration of plastid phylogenomic conflict yields new insights into the deep relationships of Leguminosae. Systematic Biology](https://academic.oup.com/sysbio/article/69/4/613/5739458)

* **Cluster_trees_by_bipartitions.py** was customized for [Zhang R, Wang YH, Jin JJ, et al. 2020. Exploration of plastid phylogenomic conflict yields new insights into the deep relationships of Leguminosae. Systematic Biology](https://academic.oup.com/sysbio/article/69/4/613/5739458)

Some scripts were deprecated:

* **plastome_arch_info.py** was deprecated here and the alternative one was merged into [GetOrganelle](https://github.com/Kinggerm/GetOrganelle).
